### PR TITLE
WIP - Issue #3631 - fix for Double data type when using MySQLi with languages that do not use '.' for the decimal point

### PR DESCRIPTION
**Pull request by mmucklo**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #3631 

#### Summary

   * Add a special ParameterType for Double
   * Detect floats in MysqliStatement and pass the 'd' as the parameter type (see: [mysqli_stmt::bind-param](https://www.php.net/manual/en/mysqli-stmt.bind-param.php#types))
   * Create two tests cases, one for the standard '.' version, and one that extends the same test case, but switches the language to de_DE and tests the ',' case.

### CODE REVIEW

File: lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php [hunk 1]
```diff
@@ -272,10 +274,17 @@ private function sendLongData($streams)
     private function bindUntypedValues(array $values)
     {
         $params = [];
-        $types  = str_repeat('s', count($values));
-
+        $types  = '';
         foreach ($values as &$v) {
             $params[] =& $v;
+            // fix for issue #3631 - detect parameter types as they have to be bound differently
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php]: Why would we want to autodetect the type of untyped parameters? They should be bound as strings by design.

File: lib/Doctrine/DBAL/Types/FloatType.php [hunk 2]
```diff
@@ -29,4 +30,12 @@ public function convertToPHPValue($value, AbstractPlatform $platform)
     {
         return $value === null ? null : (float) $value;
     }
+
+    /**
+     * {@inheritdoc}
```
Comment (morozov) [inline lib/Doctrine/DBAL/Types/FloatType.php]: Is this necessary?

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        $table->addColumn('val', 'float');
+        $table->setPrimaryKey(['id']);
+
+        $sm = $this->connection->getSchemaManager();
+        $sm->dropAndCreateTable($table);
+    }
+
+    public function testInsertAndSelect() : void
+    {
+        $value1 = 1.1;
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: If these values are independent, they should be injected using a data provider.

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        $value2 = 77.99999999999;
+        $value3 = microtime(true);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: How are these three values different? Do they test different code paths?

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            $this->fail('Expected $result3 to not be false');
+
+            return;
+        }
+
+        $diff1 = abs($result1 - $value1);
+        $diff2 = abs($result2 - $value2);
+        $diff3 = abs($result3 - $value3);
+
+        $this->assertLessThanOrEqual(0.0001, $diff1, sprintf('%f, %f, %f', $diff1, $result1, $value1));
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: Where does `0.001` come from?

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        $this->assertLessThanOrEqual(0.0001, $diff2, sprintf('%f, %f, %f', $diff2, $result2, $value2));
+        $this->assertLessThanOrEqual(0.0001, $diff3, sprintf('%f, %f, %f', $diff3, $result3, $value3));
+
+        $result1 = $this->selectDouble($value1);
+        $result2 = $this->selectDouble($value2);
+        $result3 = $this->selectDouble($value3);
+
+        $this->assertSame(is_int($result1) ? 1 : '1', $result1);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: There shouldn't be any conditionals in a test.

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * @return mixed
+     */
+    private function select(int $id)
+    {
+        return $this->connection->fetchColumn(
+            'SELECT val FROM double_table WHERE id = ?',
+            [$id],
+            0,
+            [ParameterType::INTEGER]
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: How is this relevant to the test? We're testing binding double values, not integers.

File: tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\DBAL\Functional\Types;
+
+use const LC_ALL;
+use function setlocale;
+
+class LanguageDoubleTest extends DoubleTest
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php]: Please avoid extending tests if possible. Specifically in this case, if you want to test the same code paths using different locales, use a data provider.

File: tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+{
+    protected function setUp() : void
+    {
+        setlocale(LC_ALL, 'de_DE.UTF-8', 'de_DE', 'de', 'ge');
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php]: What if neither of the locales is available at the runtime? Should there be any teardown logic in the test?

**Review by morozov — CHANGES_REQUESTED** [review]

File: tests/travis/install-language-de.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+set -ex
+
+echo "Installing language pack de..."
+
+sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 6B05F25D762E3157
```
[the remark this answers is not in the document]
Comment (mmucklo) [reply tests/travis/install-language-de.sh]: @morozov 

apt update (the next command) was failing before without adding that key, appears to be fixed.

Removed work around.

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+        $this->insert(1, $value1);
+        $this->insert(2, $value2);
+        $this->insert(3, $value3);
+
+        $result1 = $this->select(1);
+        $result2 = $this->select(2);
+        $result3 = $this->select(3);
+
+        if (is_string($result1)) {
```
[the remark this answers is not in the document]
Comment (mmucklo) [reply tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: @morozov 

Not all SQL drivers operate the same as I discovered while testing this.

PDO in most if not all cases seems to return floats as strings, while mysqli appears to return them as actual floats.

I discovered this while testing.

As I wanted the test to work on all drivers, I need to convert the ones that are strings into floats so that my comparisons properly work.

This could probably be done differently, I'm open to suggestions.

File: lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php (hunk 1 above)
Comment (mmucklo) [reply lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php]: @morozov,

Removed this for now (was missing the case for double / d anyway, as I was still testing things out) - this is for a separate "part II" of the bug where another test fails when under de_DE when dealing with floats.

I may add it back later, or do a follow on PR once this is accepted.

File: lib/Doctrine/DBAL/Types/FloatType.php (hunk 2 above)
Comment (mmucklo) [reply lib/Doctrine/DBAL/Types/FloatType.php]: @morozov 

Just following convention set in IntegerType - do you want me to remove the inheritdoc?:

[https://github.com/doctrine/dbal/blob/9ff47e738575ab208f3ab4031412c58eb676ed1e/lib/Doctrine/DBAL/Types/IntegerType.php#L38](https://github.com/doctrine/dbal/blob/9ff47e738575ab208f3ab4031412c58eb676ed1e/lib/Doctrine/DBAL/Types/IntegerType.php#L38)

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php (hunk 6 above)
Comment (mmucklo) [reply tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: @morozov 

Similar to my other comment, some drivers seem to return integers as strings in which case I have to either cast everything, or just detect.

I'm open to alternatives?

File: tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php (hunk 7 above)
Comment (mmucklo) [reply tests/Doctrine/Tests/DBAL/Functional/Types/DoubleTest.php]: @morozov 

As this is a test with Doubles, it's trying to test both insertion and result sets of the doubles to make sure what was inserted matches what's queried.

I can just test insertions / updates if you want things more concise. Although without querying the database, I may not know for sure if my insertion actually worked.

Please advise?

File: tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php (hunk 8 above)
Comment (mmucklo) [reply tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php]: @morozov 

Let me look into it.

File: tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php (hunk 9 above)
Comment (mmucklo) [reply tests/Doctrine/Tests/DBAL/Functional/Types/LanguageDoubleTest.php]: @morozov 

I can make a tear down.

For travis, every run has the de_DE language installed, but I suppose I could try to add detection logic here if it works.

**Comment by morozov** [comment]
Closing as stale.
